frontend/src/component/ui/Modal.spec.tsx

```diff
@@ -0,0 +1,64 @@
+import React from 'react';
+import { render, fireEvent } from '@testing-library/react';
+import { Modal } from './Modal';
+
+jest.useFakeTimers();
+
+it('renders without crashing', () => {
+  render(<Modal />);
+});
+
+it('opens correctly', () => {
+  const modalRef = React.createRef<Modal>();
+  render(<Modal ref={modalRef} canClose={true} />);
+
+  expect(modalRef).toBeTruthy();
+  expect(modalRef.current).toBeTruthy();
+
+  modalRef.current!.open();
+
+  expect(modalRef.current!.state.isOpen).toBe(true);
+  jest.runAllTimers();
+  expect(modalRef.current!.state.isShowing).toBe(true);
+});
+
+it('closes modal when click on close button', () => {
+  const modalRef = React.createRef<Modal>();
+  render(<Modal ref={modalRef} canClose={true} />);
+
+  modalRef.current!.open();
+  jest.runAllTimers();
+
+  modalRef.current!.close();
+
+  expect(modalRef.current!.state.isShowing).toBe(false);
+  jest.runAllTimers();
+  expect(modalRef.current!.state.isOpen).toBe(false);
+});
+
+it('closes when click on backdrop', () => {
+  const modalRef = React.createRef<Modal>();
+  const { container } = render(<Modal ref={modalRef} canClose={true} />);
+
+  modalRef.current!.open();
+  jest.runAllTimers();
+
```

Comment:
```suggestion

expect(modalRef.current!.state.isShowing).toBe(true);
expect(modalRef.current!.state.isOpen).toBe(true);
```
We need to ensure the modal was open before so that we are confident that the state actually changes

Reply:
I wonder if there is a way to reduce duplicate code, i cannot find one. (in "opens correctly" and "closes when click on backdrop")

Reply:
@kapillamba4 Well, this duplicated code is acceptable. We want to make the tests easy to reason about. Each test should fail for their own reason. When there are too many duplications, we may extract them into helpers

Reply:
Hi @kapillamba4, let's check the state before and after calling: `modalRef.current!.open()`